README
				  Gnome2::VFS
				 -------------

This module allows you to use the GNOME Virtual File System library
(libgnomevfs for short) from Perl.

The canonical API documentation can be found at

  http://developer.gnome.org/doc/API/2.0/gnome-vfs-2.0/index.html

Gnome2::VFS also comes with automatically generated API documentation.  To
access its index, use:

  perldoc Gnome2::VFS::index

				 Installation
				--------------

To install Gnome2::VFS, run the following commands:

  perl Makefile.PL
  make
  make test
  make install

				 Dependencies
				--------------

Gnome2::VFS depends on:

  ExtUtils::Depends >= 0.20
  ExtUtils::PkgConfig >= 1.03
  Glib >= 1.100
  libgnomevfs >= 2.0.0


NEWS
Overview of changes in Gnome2-VFS 1.082
=======================================

* Avoid misusing the macro PL_na, thus preventing issues when Gnome2::VFS is
  used in conjunction with certain XS modules, among them XML::Parser and
  String::Approx.

Overview of changes in Gnome2-VFS 1.081
=======================================

* Fix the GnomeVFSMimeApplication handling on libgnomevfs >= 2.10.

Overview of changes in Gnome2-VFS 1.080
=======================================

 Since 1.06x (the previous stable series)
 ----------------------------------------
  * Add Gnome2::VFS::URI::resolve_symbolic_link and
    Gnome2::VFS::Drive::needs_eject.

 Since 1.070
 -----------
  * Nothing.

Overview of changes in Gnome2-VFS 1.070
=======================================

* Add Gnome2::VFS::URI::resolve_symbolic_link.
* Add Gnome2::VFS::Drive::needs_eject.

Overview of changes in Gnome2-VFS 1.061
=======================================

* Fix a test failure.
* Fix a POD error that caused the main documentation in the Gnome2::VFS POD
  page to be shadowed.

Overview of changes in Gnome2-VFS 1.060
=======================================

 Since 1.04x (the previous stable series)
 ----------------------------------------
  * Add Gnome2::VFS::Address::equal, Gnome2::VFS::Address::match and
  Gnome2::VFS::get_slow_mime_type.
  * Add Gnome2::VFS::get_mime_type_for_name and
  Gnome2::VFS::get_mime_type_for_name_and_data.

 Since 1.031
 -----------
  * Make the Gnome2::VFS::Xfer test more robust.

Overview of changes in Gnome2-VFS 1.051
=======================================

* Add Gnome2::VFS::get_mime_type_for_name and
  Gnome2::VFS::get_mime_type_for_name_and_data.

Overview of changes in Gnome2-VFS 1.050
=======================================

* Fix a test suite issue.
* Add Gnome2::VFS::Address::equal, Gnome2::VFS::Address::match and
  Gnome2::VFS::get_slow_mime_type.

Overview of changes in Gnome2-VFS 1.041
=======================================

* Get rid of the Gtk2 dependency.

Overview of changes in Gnome2-VFS 1.040
=======================================

 Since 1.02x (the previous stable series)
 ----------------------------------------
  * Export some constants on demand.
  * Add new API in Gnome2::VFS::Mime::Type, Gnome2::VFS::Mime::Application,
    Gnome2::VFS::Handle and Gnome2::VFS.
  * Fix Gnome2::VFS::make_uri_from_input_with_trailing_ws by adding the missing
    class parameter.

 Since 1.031
 -----------
  * Fix a test suite issue.

Overview of changes in Gnome2-VFS 1.031
=======================================

* Add an example demonstrating Gnome2::VFS::Monitor.
* Fix minor test suite issues.
* Fix Gnome2::VFS::make_uri_from_input_with_trailing_ws by adding the missing
  class parameter.

Overview of changes in Gnome2-VFS 1.030
=======================================

* Export the following constants on demand GNOME_VFS_PRIORITY_MIN,
  GNOME_VFS_PRIORITY_MAX, GNOME_VFS_PRIORITY_DEFAULT,
  GNOME_VFS_SIZE_FORMAT_STR, GNOME_VFS_OFFSET_FORMAT_STR,
  GNOME_VFS_MIME_TYPE_UNKNOWN, GNOME_VFS_URI_MAGIC_STR, and
  GNOME_VFS_URI_PATH_STR.

* Add Gnome2::VFS::Mime::Type::get_default_application_for_uri,
  Gnome2::VFS::Mime::Type::get_all_applications_for_uri,
  Gnome2::VFS::Mime::Application::new_from_desktop_id,
  Gnome2::VFS::Mime::Application::get_desktop_id,
  Gnome2::VFS::Mime::Application::get_desktop_file_path,
  Gnome2::VFS::Mime::Application::get_name,
  Gnome2::VFS::Mime::Application::get_generic_name,
  Gnome2::VFS::Mime::Application::get_icon,
  Gnome2::VFS::Mime::Application::get_exec,
  Gnome2::VFS::Mime::Application::get_binary_name,
  Gnome2::VFS::Mime::Application::supports_uris,
  Gnome2::VFS::Mime::Application::requires_terminal,
  Gnome2::VFS::Mime::Application::supports_startup_notification and
  Gnome2::VFS::Mime::Application::get_startup_wm_class,
  Gnome2::VFS::Handle::forget_cache and
  Gnome2::VFS::make_uri_from_input_with_trailing_ws.

Overview of changes in Gnome2-VFS 1.022
=======================================

* Add a monitor example.  [Christian Borup]
* Fix a testsuite issue.

Overview of changes in Gnome2-VFS 1.021
=======================================

* Use simple opaque scalars instead of boxed types for several objects.
* Some internal reorganization.
* Documentation fixes.

Overview of changes in Gnome2-VFS 1.020
=======================================

 Since 1.00x (the previous stable series)
 ----------------------------------------
  * New objects: Gnome2::VFS::Address, Gnome2::VFS::DNSSD, Gnome2::VFS::Drive,
    Gnome2::VFS::Resolve, Gnome2::VFS::Volume, and Gnome2::VFS::VolumeMonitor.
  * New API in: Gnome2::VFS::Mime, Gnome2::VFS::Async, and
    Gnome2::VFS::FileInfo.
  * Plug a few leaks.

 Since 1.011
 -----------
  * "Fix" a test failure.

Overview of changes in Gnome2-VFS 1.011
=======================================

* Don't leak the GnomeVFSDrive and GnomeVFSVolume callbacks.

Overview of changes in Gnome2-VFS 1.010
=======================================

* Depend on Glib 1.053.
* Wrapped GnomeVFSAddress, GnomeVFSDNSSD, GnomeVFSDrive, GnomeVFSResolve,
  GnomeVFSVolume, and GnomeVFSVolumeMonitor.
* Wrapped new stuff in GnomeVFSMime and GnomeVFSAsync.
* Implemented a constructor for GnomeVFSFileInfo.

Overview of changes in Gnome2-VFS 1.003
=======================================

* Make Gnome2::VFS::read_entire_file not crash on invalid input.
* Fix a small test bug.
* Don't depend on Gtk2 at runtime.

Overview of changes in Gnome2-VFS 1.002
=======================================

* Hushed compiler warnings that might occur under certain circumstances.

Overview of changes in Gnome2-VFS 1.001
=======================================

* Made all version checks refer to stable releases.
* Made the test suite more robust.

Overview of changes in Gnome2-VFS 1.00
======================================

* Depend on Glib 1.040.

Overview of changes in Gnome2-VFS 0.99
======================================


vfs2perl.h
/*
 * Copyright (C) 2003 by the gtk2-perl team
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Id$
 */

#ifndef _VFS2PERL_H_
#define _VFS2PERL_H_

#include <gperl.h>
#include "vfs2perl-version.h"

#include <libgnomevfs/gnome-vfs.h>
#include <libgnomevfs/gnome-vfs-uri.h>
#include <libgnomevfs/gnome-vfs-handle.h>
#include <libgnomevfs/gnome-vfs-mime-handlers.h>
#include <libgnomevfs/gnome-vfs-mime-monitor.h>
#include <libgnomevfs/gnome-vfs-application-registry.h>

#if VFS_CHECK_VERSION (2, 8, 0)

#include <libgnomevfs/gnome-vfs-dns-sd.h>
#include <libgnomevfs/gnome-vfs-address.h>
#include <libgnomevfs/gnome-vfs-resolve.h>

#endif /* 2.8 */

/* ------------------------------------------------------------------------- */

#define GNOME_VFS_TYPE_VFS_URI (vfs2perl_gnome_vfs_uri_get_type ())
GType vfs2perl_gnome_vfs_uri_get_type (void) G_GNUC_CONST;

/* ------------------------------------------------------------------------- */

#include "vfs2perl-gtypes.h"

/* i'm just guessing here.  if you get a message about failed assertions
 * that something is a GFlags type in GnomeVFSDirectory.t, then you probably
 * need to set this to include your version.
 */
#if !VFS_CHECK_VERSION (2, 1, 0)
# define VFS2PERL_BROKEN_FILE_PERMISSIONS
# undef GNOME_VFS_TYPE_VFS_FILE_PERMISSIONS
# define GNOME_VFS_TYPE_VFS_FILE_PERMISSIONS (_vfs2perl_gnome_vfs_file_permissions_get_type ())
  GType _vfs2perl_gnome_vfs_file_permissions_get_type (void) G_GNUC_CONST;
#endif

#include "vfs2perl-autogen.h"

/* ------------------------------------------------------------------------- */

SV * newSVGnomeVFSFileSize (GnomeVFSFileSize size);
GnomeVFSFileSize SvGnomeVFSFileSize (SV *size);

SV * newSVGnomeVFSFileOffset (GnomeVFSFileOffset offset);
GnomeVFSFileOffset SvGnomeVFSFileOffset (SV *offset);

SV * newSVGnomeVFSHandle (GnomeVFSHandle *handle);
GnomeVFSHandle * SvGnomeVFSHandle (SV *handle);

SV * newSVGnomeVFSMonitorHandle (GnomeVFSMonitorHandle *handle);
GnomeVFSMonitorHandle * SvGnomeVFSMonitorHandle (SV *handle);

SV * newSVGnomeVFSDirectoryHandle (GnomeVFSDirectoryHandle *handle);
GnomeVFSDirectoryHandle * SvGnomeVFSDirectoryHandle (SV *handle);

SV * newSVGnomeVFSAsyncHandle (GnomeVFSAsyncHandle *handle);
GnomeVFSAsyncHandle * SvGnomeVFSAsyncHandle (SV *handle);

#if VFS_CHECK_VERSION (2, 8, 0)

SV * newSVGnomeVFSDNSSDBrowseHandle (GnomeVFSDNSSDBrowseHandle *handle);
GnomeVFSDNSSDBrowseHandle * SvGnomeVFSDNSSDBrowseHandle (SV *handle);

SV * newSVGnomeVFSDNSSDResolveHandle (GnomeVFSDNSSDResolveHandle *handle);
GnomeVFSDNSSDResolveHandle * SvGnomeVFSDNSSDResolveHandle (SV *handle);

SV * newSVGnomeVFSResolveHandle (GnomeVFSResolveHandle *handle);
GnomeVFSResolveHandle * SvGnomeVFSResolveHandle (SV *handle);

#endif /* 2.8 */

typedef const char GnomeVFSApplication;
const char * SvGnomeVFSApplication (SV *object);
SV * newSVGnomeVFSApplication (const char *app_id);

typedef const char GnomeVFSMimeType;
const char * SvGnomeVFSMimeType (SV *object);
SV * newSVGnomeVFSMimeType (const char *mime_type);

SV * newSVGnomeVFSMimeApplication (GnomeVFSMimeApplication *application);
GnomeVFSMimeApplication * SvGnomeVFSMimeApplication (SV *object);

GnomeVFSFileInfo * SvGnomeVFSFileInfo (SV *object);
SV * newSVGnomeVFSFileInfo (GnomeVFSFileInfo *info);

SV * newSVGnomeVFSXferProgressInfo (GnomeVFSXferProgressInfo *info);

/* ------------------------------------------------------------------------- */

GList * SvPVGList (SV *ref);

GList * SvGnomeVFSURIGList (SV *ref);

char ** SvEnvArray (SV *ref);

SV * newSVGnomeVFSFileInfoGList (GList *list);

SV * newSVGnomeVFSGetFileInfoResultGList (GList *list);

SV * newSVGnomeVFSFindDirectoryResultGList (GList *list);

/* ------------------------------------------------------------------------- */

#endif /* _VFS2PERL_H_ */


Makefile.PL
#!/usr/bin/perl -w
use 5.008;
use strict;
use ExtUtils::MakeMaker;
use Cwd;
use File::Spec;

# $Header: /cvsroot/gtk2-perl/gtk2-perl-xs/Gnome2-VFS/Makefile.PL,v 1.29 2006/03/12 22:08:19 kaffeetisch Exp $

our %build_reqs = (
  'perl-ExtUtils-Depends'   => '0.20',
  'perl-ExtUtils-PkgConfig' => '1.03',
  'perl-Glib'               => '1.120',
  'GnomeVFS'                => '2.0.0',
);

our %PREREQ_PM = (
  "Glib"                => $build_reqs{'perl-Glib'},
  "ExtUtils::Depends"   => $build_reqs{'perl-ExtUtils-Depends'},
  "ExtUtils::PkgConfig" => $build_reqs{'perl-ExtUtils-PkgConfig'},
);

unless (eval "use ExtUtils::Depends '$build_reqs{'perl-ExtUtils-Depends'}';"
           . "use ExtUtils::PkgConfig '$build_reqs{'perl-ExtUtils-PkgConfig'}';"
           . "use Glib '$build_reqs{'perl-Glib'}';"
           . "use Glib::MakeHelper;"
	   . "use Glib::CodeGen;"
           . "1") {
  warn "$@\n";
  WriteMakefile(
    PREREQ_FATAL => 1,
    PREREQ_PM    => \%PREREQ_PM,
  );
  exit 1; # not reached
}

my %vfs_pkgcfg = ExtUtils::PkgConfig->find("gnome-vfs-2.0 >= $build_reqs{GnomeVFS}");

###############################################################################

mkdir 'build', 0777;

my ($major, $minor, $micro) = split /\./, $vfs_pkgcfg{modversion};
my $needs_gtypes = ($major < 2 or
                    $major == 2 && $minor < 5 or
                    $major == 2 && $minor == 5 && $micro < 3);

if ($needs_gtypes) {
  chomp(my $vfs_includes = `pkg-config --variable includedir gnome-vfs-2.0`);
  my @vfs_headers = glob($vfs_includes . "/gnome-vfs-2.0/libgnomevfs/*.h");

  system(q(glib-mkenums --fhead "#ifndef __VFS2PERL_GTYPES_H__\n" ) .
                      q(--fhead "#define __VFS2PERL_GTYPES_H__ 1\n\n" ) .
                      q(--fhead "#include <glib-object.h>\n\n" ) .
                      q(--fhead "G_BEGIN_DECLS\n\n" ) .
                      q(--eprod "#ifndef GNOME_VFS_TYPE_@ENUMSHORT@\n" ) .
                      q(--eprod "#define GNOME_VFS_TYPE_@ENUMSHORT@ vfs2perl_@enum_name@_get_type()\n" ) .
                      q(--eprod "GType vfs2perl_@enum_name@_get_type (void);\n" ) .
                      q(--eprod "#endif /* !GNOME_VFS_TYPE_@ENUMSHORT@ */\n" ) .
                      q(--ftail "G_END_DECLS\n\n" ) .
                      q(--ftail "#endif /* __VFS2PERL_GTYPES_H__ */\n" ) .
                      "@vfs_headers > build/vfs2perl-gtypes.h");

  system(q(glib-mkenums --fhead "#include <libgnomevfs/gnome-vfs-types.h>\n" ) .
                      q(--fhead "#include <libgnomevfs/gnome-vfs-utils.h>\n" ) .
                      q(--fhead "#include <libgnomevfs/gnome-vfs-monitor.h>\n" ) .
                      q(--fhead "#include <libgnomevfs/gnome-vfs-mime-handlers.h>\n" ) .
                      q(--fhead "#include <libgnomevfs/gnome-vfs-directory.h>\n" ) .
                      q(--vhead "static const G@Type@Value _vfs2perl_@enum_name@_values[] = {" ) .
                      q(--vprod "  { @VALUENAME@, \"@VALUENAME@\", \"@valuenick@\" }," ) .
                      q(--vtail "  { 0, NULL, NULL }\n};\n\n" ) .
                      q(--vtail "GType\nvfs2perl_@enum_name@_get_type (void)\n{\n" ) .
                      q(--vtail "  static GType type = 0;\n\n" ) .
                      q(--vtail "  if (!type)\n" ) .
                      q(--vtail "    type = g_@type@_register_static (\"@EnumName@\", _vfs2perl_@enum_name@_values);\n\n" ) .
                      q(--vtail "  return type;\n}\n\n" ) .
                      "@vfs_headers > build/vfs2perl-gtypes.c");

  # UGLY, UGLY, UGLY!
  system("perl -pi -e 's/GNOME_VFS_TYPE_VFSURI_HIDE_OPTIONS/GNOME_VFS_TYPE_VFS_URI_HIDE_OPTIONS/g;
	               s/vfs2perl_gnome_vfsuri_hide_options/vfs2perl_gnome_vfs_uri_hide_options/g' build/vfs2perl-gtypes.h build/vfs2perl-gtypes.c");
}
else {
  open GTYPES, ">build/vfs2perl-gtypes.h" or die("Opening '>build/vfs2perl-gtypes.h': $!");
  print GTYPES "#include <libgnomevfs/gnome-vfs-enum-types.h>\n";
  close GTYPES;
}

###############################################################################

my @xs_files = ();

foreach (Glib::MakeHelper->select_files_by_version ("xs_files", split /\./, $vfs_pkgcfg{modversion})) {
	push @xs_files, Glib::MakeHelper->read_source_list_file ($_);
}

my %pod_files = (
  'VFS.pm' => '$(INST_MAN3DIR)/Gnome2::VFS.$(MAN3EXT)',
  Glib::MakeHelper->do_pod_files (@xs_files),
);

Glib::CodeGen->parse_maps ('vfs2perl', input => [<maps-*\.*>]);
Glib::CodeGen->write_boot (xs_files => \@xs_files,
                           ignore => qr/^Gnome2::VFS$/);

ExtUtils::PkgConfig->write_version_macros (
  "build/vfs2perl-version.h",
  "gnome-vfs-2.0" => "VFS",
);

###############################################################################

my $vfs = ExtUtils::Depends->new ('Gnome2::VFS', 'Glib');
$vfs->set_inc ($vfs_pkgcfg{cflags} . " -I build");
$vfs->set_libs ($vfs_pkgcfg{libs});
$vfs->add_xs (@xs_files);
$vfs->add_c ("vfs2perl.c");
$vfs->add_c ("build/vfs2perl-gtypes.c") if ($needs_gtypes);
$vfs->add_pm ('VFS.pm' => '$(INST_LIBDIR)/VFS.pm');
$vfs->add_typemaps (map {File::Spec->catfile(cwd(), $_)} 'build/vfs2perl.typemap', 'vfs.typemap');

$vfs->install (qw(vfs2perl.h
                  build/vfs2perl-autogen.h
                  build/vfs2perl-version.h
                  build/vfs2perl-gtypes.h
                  doctypes));
$vfs->save_config ('build/IFiles.pm');

WriteMakefile(
  NAME            => 'Gnome2::VFS',
  VERSION_FROM    => 'VFS.pm', # finds $VERSION
  ABSTRACT_FROM   => 'VFS.pm', # retrieve abstract from module
  PREREQ_PM       => \%PREREQ_PM,
  XSPROTOARG      => '-noprototypes',
  MAN3PODS        => \%pod_files,
  $vfs->get_makefile_vars,
);

###############################################################################

sub MY::postamble {
  return Glib::MakeHelper->postamble_clean ()
       . Glib::MakeHelper->postamble_docs_full (
		DEPENDS => $vfs,
		DOCTYPES => 'doctypes',
		COPYRIGHT_FROM => 'copyright.pod'
         )
       . Glib::MakeHelper->postamble_rpms (
		'GNOME_VFS' => $build_reqs{'GnomeVFS'},
		'PERL_EXTUTILS_DEPENDS' => 
			$build_reqs{'perl-ExtUtils-Depends'},
		'PERL_EXTUTILS_PKGCONFIG' => 
			$build_reqs{'perl-ExtUtils-PkgConfig'},
		'PERL_GLIB' => $build_reqs{'perl-Glib'},
	);
}


VFS.pm
# $Id$

package Gnome2::VFS;

use 5.008;
use strict;
use warnings;

use Glib;

require Exporter;
require DynaLoader;

our @ISA = qw(Exporter DynaLoader);

our @EXPORT_OK = qw(
  GNOME_VFS_PRIORITY_MIN
  GNOME_VFS_PRIORITY_MAX
  GNOME_VFS_PRIORITY_DEFAULT
  GNOME_VFS_SIZE_FORMAT_STR
  GNOME_VFS_OFFSET_FORMAT_STR
  GNOME_VFS_MIME_TYPE_UNKNOWN
  GNOME_VFS_URI_MAGIC_STR
  GNOME_VFS_URI_PATH_STR
);

# --------------------------------------------------------------------------- #

our $VERSION = '1.082';

sub import {
  my ($self) = @_;
  my @symbols = ();

  foreach (@_) {
    if (/^-?init$/) {
      $self -> init();
    } else {
      push @symbols, $_;
    }
  }

  Gnome2::VFS -> export_to_level(1, @symbols);
}

sub dl_load_flags { $^O eq 'darwin' ? 0x00 : 0x01 }

Gnome2::VFS -> bootstrap($VERSION);

# --------------------------------------------------------------------------- #

use constant GNOME_VFS_PRIORITY_MIN => -10;
use constant GNOME_VFS_PRIORITY_MAX => 10;
use constant GNOME_VFS_PRIORITY_DEFAULT => 0;

use constant GNOME_VFS_SIZE_FORMAT_STR => "Lu";
use constant GNOME_VFS_OFFSET_FORMAT_STR => "Ld";

use constant GNOME_VFS_MIME_TYPE_UNKNOWN => "application/octet-stream";

use constant GNOME_VFS_URI_MAGIC_STR => "#";
use constant GNOME_VFS_URI_PATH_STR => "/";

1;

# --------------------------------------------------------------------------- #

__END__

=head1 NAME

Gnome2::VFS - Perl interface to the 2.x series of the GNOME VFS library

=head1 SYNOPSIS

  use Gnome2::VFS;

  sub die_already {
    my ($action) = @_;
    die("An error occured while $action.\n");
  }

  die_already("initializing GNOME VFS") unless (Gnome2::VFS -> init());

  my $source = "http://www.perldoc.com/about.html";
  my ($result, $handle, $info);

  # Open a connection to Perldoc.
  ($result, $handle) = Gnome2::VFS -> open($source, "read");
  die_already("opening connection to '$source'")
    unless ($result eq "ok");

  # Get the file information.
  ($result, $info) = $handle -> get_file_info("default");
  die_already("retrieving information about '$source'")
    unless ($result eq "ok");

  # Read the content.
  my $bytes = $info -> { size };

  my $bytes_read = 0;
  my $buffer = "";

  do {
    my ($tmp_buffer, $tmp_bytes_read);

    ($result, $tmp_bytes_read, $tmp_buffer) =
      $handle -> read($bytes - $bytes_read);

    $buffer .= $tmp_buffer;
    $bytes_read += $tmp_bytes_read;
  } while ($result eq "ok" and $bytes_read < $bytes);

  die_already("reading $bytes bytes from '$source'")
    unless ($result eq "ok" && $bytes_read == $bytes);

  # Close the connection.
  $result = $handle -> close();
  die_already("closing connection to '$source'")
    unless ($result eq "ok");

  # Create and open the target.
  my $target = "/tmp/" . $info -> { name };
  my $uri = Gnome2::VFS::URI -> new($target);

  ($result, $handle) = $uri -> create("write", 1, 0644);
  die_already("creating '$target'") unless ($result eq "ok");

  # Write to it.
  my $bytes_written;

  ($result, $bytes_written) = $handle -> write($buffer, $bytes);
  die_already("writing $bytes bytes to '$target'")
    unless ($result eq "ok" && $bytes_written == $bytes);

  # Close the target.
  $result = $handle -> close();
  die_already("closing '$target'") unless ($result eq "ok");

  Gnome2::VFS -> shutdown();

=head1 ABSTRACT

This module allows you to interface with the GNOME Virtual File System library.
It provides the means to transparently access files on all kinds of
filesystems.

=head1 DESCRIPTION

Since this module tries to stick very closely to the C API, the documentation
found at

  L<http://developer.gnome.org/doc/API/2.0/gnome-vfs-2.0/>

is the canonical reference.

In addition to that, there's also the automatically generated API
documentation: L<Gnome2::VFS::index>.

The mapping described in L<Gtk2::api> also applies to this module.

To discuss this module, ask questions and flame/praise the authors, join
gtk-perl-list@gnome.org at lists.gnome.org.

=head1 KNOWN BUGS

There are some memory leaks especially with respect to callbacks.  This mainly
affects GnomeVFSAsync as well as some parts of GnomeVFSXfer and GnomeVFSOps.
GnomeVFSMime leaks some list data.

GnomeVFSAsync is also known to crash under certain conditions when there are
many concurrent transfers.

=head1 SEE ALSO

L<Gnome2::VFS::index>, L<Glib>, L<Gtk2>, L<Gtk2::api>.

=head1 AUTHOR

Torsten Schoenfeld E<lt>kaffeetisch@web.deE<gt>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003-2007 by the gtk2-perl team (see the file AUTHORS)

This library is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 2.1 of the License, or (at your option) any
later version.

This library is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU Lesser General Public License along
with this library; if not, write to the Free Software Foundation, Inc.,
59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

=cut
